File: ios/Capacitor/Capacitor/Plugins/Permissions.swift
import Foundation
import Photos
import UserNotifications

/**
 * Allow to check hardware permissions
 */
@objc(CAPPermissionsPlugin)
public class CAPPermissionsPlugin: CAPPlugin {
  fileprivate static let uknownPermissionValue = "Unknown permission value"
  
  @objc func query(_ call: CAPPluginCall) {
    guard let name = call.getString("name") else {
      call.reject("Must provide a permission to check")
      return
    }
    
    switch (name) {
    case "camera":
      return checkCamera(call)
    case "geolocation":
      return checkGeolocation(call)
    case "notifications":
      return checkNotifications(call)
    case "clipboard-read", "clipboard-write":
      return checkClipboard(call)
    case "photos":
      return checkPhotos(call)
    case "microphone":
      return checkMicrophone(call)
    default:
      return call.reject("Unknown permission type")
    }
  }

  func checkCamera(_ call: CAPPluginCall) {
    let authStatus = AVCaptureDevice.authorizationStatus(for: .video)
    
    let ret: String
    switch (authStatus) {
    case .denied, .restricted:
      ret = "denied"
    case .authorized:
      ret = "granted"
    case .notDetermined:
      ret = "prompt"
    @unknown default:
      call.reject(CAPPermissionsPlugin.uknownPermissionValue)
      return
    }

    call.resolve([
      "state": ret
    ])
  }

  func checkPhotos(_ call: CAPPluginCall) {
    let photoAuthorizationStatus = PHPhotoLibrary.authorizationStatus()
    
    let ret: String
    switch (photoAuthorizationStatus) {
    case .denied, .restricted:
      ret = "denied"
    case .authorized:
      ret = "granted"
    #if swift(>=5.3)
    case .limited:
      // TODO: address this new case properly
      #warning(".limited != .authorized, authorization status should be revisited for iOS 14")
      ret = "granted"
    #endif
    case .notDetermined:
      ret = "prompt"
    @unknown default:
      call.reject(CAPPermissionsPlugin.uknownPermissionValue)
      return
    }
    
    call.resolve([
      "state": ret
    ])
  }

  func checkGeolocation(_ call: CAPPluginCall) {
    let ret: String
    if CLLocationManager.locationServicesEnabled() {
        switch CLLocationManager.authorizationStatus() {
        case .denied, .restricted:
          ret = "denied"
        case .authorizedAlways, .authorizedWhenInUse:
          ret = "granted"
        case .notDetermined:
          ret = "prompt"
        @unknown default:
          call.reject(CAPPermissionsPlugin.uknownPermissionValue)
          return
        }
    } else {
      ret = "denied"
    }

    call.resolve([
      "state": ret
    ])
  }

  func checkNotifications(_ call: CAPPluginCall) {
    UNUserNotificationCenter.current().getNotificationSettings(completionHandler: { settings in
      let ret: String
      switch settings.authorizationStatus {
      case .authorized, .provisional:
        ret = "granted"
      #if swift(>=5.3)
      case .ephemeral:
        ret = "granted"
      #endif
      case .denied:
        ret = "denied"
      case .notDetermined:
        ret = "prompt"
      @unknown default:
        call.reject(CAPPermissionsPlugin.uknownPermissionValue)
        return
      }
      
      call.resolve([
        "state": ret
      ])
    })
  }
  
  func checkClipboard(_ call: CAPPluginCall) {
    call.resolve([
      "state": "granted"
    ])
  }

  func checkMicrophone(_ call: CAPPluginCall) {
    let microStatus = AVCaptureDevice.authorizationStatus(for: .audio)

    let ret: String
    switch (microStatus) {
    case .authorized:
      ret = "granted"
    case .denied, .restricted:
      ret = "denied"
    case .notDetermined:
      ret = "prompt"
    @unknown default:
      call.reject(CAPPermissionsPlugin.uknownPermissionValue)
      return
    }

    call.resolve([
      "state": ret
    ])
  }
}

File: android/capacitor/src/main/java/com/getcapacitor/plugin/Permissions.java
package com.getcapacitor.plugin;

import android.Manifest;
import android.content.pm.PackageManager;

import androidx.core.app.NotificationManagerCompat;
import androidx.core.content.ContextCompat;

import com.getcapacitor.JSObject;
import com.getcapacitor.NativePlugin;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;

import org.json.JSONArray;

@NativePlugin()
public class Permissions extends Plugin {

  @PluginMethod
  public void query(PluginCall call) {
    String name = call.getString("name");

    switch (name) {
      case "camera":
        checkCamera(call);
        break;
      case "photos":
        checkPhotos(call);
        break;
      case "geolocation":
        checkGeo(call);
        break;
      case "notifications":
        checkNotifications(call);
        break;
      case "clipboard-read":
      case "clipboard-write":
        checkClipboard(call);
      case "microphone":
        checkMicrophone(call);
        break;
      default:
        call.reject("Unknown permission type");
    }
  }

  private void checkPerm(String perm, PluginCall call) {
    JSObject ret = new JSObject();
    if (ContextCompat.checkSelfPermission(getContext(), perm) == PackageManager.PERMISSION_DENIED) {
      ret.put("state", "denied");
    } else if (ContextCompat.checkSelfPermission(getContext(), perm) == PackageManager.PERMISSION_GRANTED) {
      ret.put("state", "granted");
    } else {
      ret.put("state", "prompt");
    }
    call.resolve(ret);
  }

  private void checkCamera(PluginCall call) {
    checkPerm(Manifest.permission.CAMERA, call);
  }

  private void checkPhotos(PluginCall call) {
    checkPerm(Manifest.permission.READ_EXTERNAL_STORAGE, call);
  }

  private void checkGeo(PluginCall call) {
    checkPerm(Manifest.permission.ACCESS_COARSE_LOCATION, call);
  }

  private void checkNotifications(PluginCall call) {
    boolean areEnabled = NotificationManagerCompat.from(getContext()).areNotificationsEnabled();
    JSObject ret = new JSObject();
    ret.put("state", areEnabled ? "granted" : "denied");
    call.resolve(ret);
  }

  private void checkClipboard(PluginCall call) {
    JSObject ret = new JSObject();
    ret.put("state", "granted");
    call.resolve(ret);
  }

  private void checkMicrophone(PluginCall call) {
    checkPerm(Manifest.permission.RECORD_AUDIO, call);
  }

}
